File: ios/QuailProcessor.swift
import Foundation
import AicSdk

/// Wraps an aic-sdk Quail speech-enhancement processor for the outgoing mic path.
///
/// Threading contract: `process(_:count:)` must only ever be called from the
/// audio tap thread (the SDK's process call is real-time safe but not
/// thread-safe). `setEnabled` may be called from any thread (the SDK context
/// is thread-safe). `reset()` must only be called while the tap is **not**
/// delivering buffers (i.e. after the engine is stopped or recording is off),
/// because it mutates the same `pending` accumulator that `process(_:count:)`
/// uses — concurrent mutation of a Swift Array is undefined behaviour.
final class QuailProcessor {
    static let sampleRate: UInt32 = 24000

    private var model: OpaquePointer?
    private var processor: OpaquePointer?
    private var context: OpaquePointer?
    private(set) var optimalFrameCount: Int = 0

    // Mic samples accumulate here until a full model frame is available.
    private var pending: [Float] = []
    // Latched on the first process() error within a session; cleared by reset() so each
    // new session gets at most one retry (and at most one onError event per session).
    // Written on the tap queue; read from other threads via `isAvailable`. A torn Bool
    // read is benign on arm64 and acceptable here — revisit under Swift 6 strict concurrency.
    private(set) var hasFailed = false
    var onError: ((String) -> Void)?

    var isAvailable: Bool { processor != nil && !hasFailed }

    /// Locates the bundled .aicmodel inside the AicModels resource bundle
    /// (copied into the app bundle because the pod is a static framework).
    static func bundledModelPath() -> String? {
        let candidates = [
            Bundle.main.url(forResource: "AicModels", withExtension: "bundle"),
            Bundle(for: QuailProcessor.self).url(forResource: "AicModels", withExtension: "bundle"),
        ]
        for case let url? in candidates {
            if let bundle = Bundle(url: url),
               let path = bundle.paths(forResourcesOfType: "aicmodel", inDirectory: nil).first {
                // Enumerate instead of hardcoding the filename: the vendor script controls
                // which single .aicmodel ships, keeping the two automatically in sync.
                return path
            }
        }
        return nil
    }

    init?(licenseKey: String, modelPath: String) {
        var modelHandle: OpaquePointer?
        var rc = aic_model_create_from_file(&modelHandle, modelPath)
        guard rc == AIC_ERROR_CODE_SUCCESS, let loadedModel = modelHandle else {
            NSLog("[QuailProcessor] model load failed: \(rc)")
            return nil
        }
        model = loadedModel

        var frames: Int = 0
        rc = aic_model_get_optimal_num_frames(loadedModel, Self.sampleRate, &frames)
        guard rc == AIC_ERROR_CODE_SUCCESS, frames > 0 else {
            NSLog("[QuailProcessor] optimal frame query failed: \(rc)")
            aic_model_destroy(loadedModel)
            model = nil
            return nil
        }
        optimalFrameCount = frames

        var proc: OpaquePointer?
        rc = aic_processor_create(&proc, loadedModel, licenseKey, nil)
        guard rc == AIC_ERROR_CODE_SUCCESS, let createdProc = proc else {
            NSLog("[QuailProcessor] processor create failed: \(rc)")
            aic_model_destroy(loadedModel)
            model = nil
            return nil
        }
        processor = createdProc

        rc = aic_processor_initialize(createdProc, Self.sampleRate, 1, frames, false)
        guard rc == AIC_ERROR_CODE_SUCCESS else {
            NSLog("[QuailProcessor] processor initialize failed: \(rc)")
            aic_processor_destroy(createdProc)
            aic_model_destroy(loadedModel)
            processor = nil
            model = nil
            return nil
        }

        var ctx: OpaquePointer?
        rc = aic_processor_context_create(&ctx, createdProc)
        guard rc == AIC_ERROR_CODE_SUCCESS, let createdCtx = ctx else {
            NSLog("[QuailProcessor] context create failed: \(rc)")
            aic_processor_destroy(createdProc)
            aic_model_destroy(loadedModel)
            processor = nil
            model = nil
            return nil
        }
        context = createdCtx

        pending.reserveCapacity(optimalFrameCount * 4)
    }

    deinit {
        aic_processor_context_destroy(context)
        aic_processor_destroy(processor)
        aic_model_destroy(model)
    }

    /// Enhancement toggle. Disabled uses the SDK's latency-compensated bypass
    /// (output delay unchanged), so flipping mid-stream produces no clicks.
    func setEnabled(_ enabled: Bool) {
        guard let context else { return }
        aic_processor_context_set_parameter(context, AIC_PROCESSOR_PARAMETER_BYPASS, enabled ? 0.0 : 1.0)
    }

    /// Clears SDK state and the local accumulator. Call between sessions and
    /// after audio interruptions so stale audio doesn't color the next frames.
    /// Must only be called while the audio tap is not delivering buffers (see
    /// class-level threading contract). Also resets `hasFailed` so the new
    /// session gets at most one enhancement-error retry.
    func reset() {
        if let context {
            aic_processor_context_reset(context)
        }
        pending.removeAll(keepingCapacity: true)
        hasFailed = false
    }

    /// Feeds raw mic samples; returns the enhanced samples that are ready (a
    /// multiple of the model frame size — possibly empty while accumulating).
    /// On the first SDK error this latches `hasFailed`, reports via `onError`,
    /// and returns the failing chunk as-is (possibly partially processed by the
    /// SDK) plus the remaining buffered input, so no audio is lost.
    func process(_ samples: UnsafePointer<Float>, count: Int) -> [Float] {
        if hasFailed {
            return Array(UnsafeBufferPointer(start: samples, count: count))
        }
        pending.append(contentsOf: UnsafeBufferPointer(start: samples, count: count))

        var output: [Float] = []
        // Per-callback Array allocations below are deliberate: the tap runs on a
        // non-realtime dispatch queue so heap allocation is safe. Revisit only if
        // this moves to a realtime (AudioUnit render) thread.
        while pending.count >= optimalFrameCount {
            var chunk = Array(pending.prefix(optimalFrameCount))
            pending.removeFirst(optimalFrameCount)

            let rc = chunk.withUnsafeMutableBufferPointer { buf -> AicErrorCode in
                var channel: UnsafeMutablePointer<Float>? = buf.baseAddress
                return withUnsafeMutablePointer(to: &channel) { planar in
                    aic_processor_process_planar(processor, planar, 1, optimalFrameCount)
                }
            }
            guard rc == AIC_ERROR_CODE_SUCCESS else {
                hasFailed = true
                let leftover = pending
                pending.removeAll(keepingCapacity: false)
                onError?("aic_processor_process_planar failed: \(rc)")
                return output + chunk + leftover
            }
            output.append(contentsOf: chunk)
        }
        return output
    }
}

File: android/src/main/java/expo/modules/twowayaudio/QuailProcessor.kt
package expo.modules.twowayaudio

import android.content.Context
import android.util.Log
import java.io.File

/**
 * Wraps the aic-sdk Quail speech-enhancement processor (via the JNI shim) for
 * the outgoing mic path.
 *
 * Threading: [process] is only ever called from the AudioEngine mic-tap thread
 * (the SDK process call is real-time safe but not thread-safe). [setEnabled] and
 * [reset] go through the SDK's thread-safe context and the shim's mutex, so they
 * are safe to call from any thread.
 */
class QuailProcessor private constructor(private var handle: Long) {

    /**
     * Latched on the first process() error; from then on input passes through raw.
     * @Volatile publishes the latch across threads: written on the mic-tap thread
     * (process()) and cleared on the JS-bridge thread (reset()).
     */
    @Volatile
    private var hasFailed = false
    var onError: ((String) -> Unit)? = null

    val isAvailable: Boolean
        get() = handle != 0L && !hasFailed

    /**
     * Feeds raw Int16-LE PCM mic bytes; returns the enhanced Int16-LE bytes that
     * are ready — a multiple of the model frame size, possibly EMPTY while still
     * accumulating toward a full frame. On the first SDK error this latches
     * [hasFailed], reports via [onError], and returns best-effort/passthrough
     * audio so nothing is dropped.
     */
    fun process(input: ByteArray): ByteArray {
        if (hasFailed || handle == 0L) return input
        // nativeProcess can return null if the JVM fails to allocate the output
        // array (native OOM); latch + pass the raw input through so audio is never
        // dropped and the non-null contract downstream holds.
        val out = nativeProcess(handle, input, input.size) ?: run {
            hasFailed = true
            onError?.invoke("nativeProcess returned null (native allocation failed)")
            return input
        }
        val rc = nativeTakeError(handle)
        if (rc != 0) {
            hasFailed = true
            onError?.invoke("aic_processor_process_planar failed: $rc")
        }
        return out
    }

    /**
     * Enhancement toggle. Disabled uses the SDK's latency-compensated bypass
     * (output delay unchanged), so flipping mid-stream produces no clicks.
     */
    fun setEnabled(enabled: Boolean) {
        if (handle != 0L) nativeSetEnabled(handle, enabled)
    }

    /**
     * Clears SDK state + the accumulator and un-latches [hasFailed] for a
     * per-session retry. Call between sessions and after interruptions.
     */
    fun reset() {
        if (handle != 0L) nativeReset(handle)
        hasFailed = false
    }

    fun destroy() {
        if (handle != 0L) {
            nativeDestroy(handle)
            handle = 0L
        }
    }

    /** Thrown when the native processor cannot be created; [code] is the SDK rc. */
    class InitError(val code: Int) : Exception("aic processor create failed: code=$code")

    companion object {
        private const val MODEL_ASSET = "quail_vf_2_1_s_16khz_5i8jb8of_v12.aicmodel"

        // Sentinel rc for "native libraries could not be loaded" (no SDK call was
        // reached), distinct from the SDK's non-negative error codes.
        private const val NATIVE_UNAVAILABLE = -1

        // We ship arm64-v8a only; on any other ABI the .so is absent and
        // loadLibrary throws UnsatisfiedLinkError. Catch it so the class still
        // initializes and invoke() can fail gracefully into the relay path
        // instead of crashing the app (the error escapes initialize()'s
        // catch(Exception) otherwise — UnsatisfiedLinkError is an Error).
        @Volatile
        private var nativeLibLoaded = false

        init {
            nativeLibLoaded = try {
                // Load the prebuilt SDK first so the shim's DT_NEEDED resolves on all
                // supported API levels (older linkers don't auto-resolve siblings).
                System.loadLibrary("aic")
                System.loadLibrary("twowayaudio_aic")
                true
            } catch (t: UnsatisfiedLinkError) {
                Log.e("QuailProcessor", "native voice-focus libraries unavailable for this ABI", t)
                false
            }
        }

        /**
         * Creates a processor, throwing [InitError] (with the SDK rc, or
         * [NATIVE_UNAVAILABLE] when the native libraries failed to load) on failure.
         */
        @Throws(InitError::class)
        operator fun invoke(licenseKey: String, modelPath: String): QuailProcessor {
            if (!nativeLibLoaded) throw InitError(NATIVE_UNAVAILABLE)
            val err = IntArray(1)
            val handle = nativeCreate(licenseKey, modelPath, err)
            if (handle == 0L) throw InitError(err[0])
            return QuailProcessor(handle)
        }

        /**
         * Copies the bundled .aicmodel from APK assets to filesDir (assets have no
         * fopen-able path) and returns the absolute path, or null on failure. The
         * copy is skipped when an identically-named file already exists; the model
         * filename carries its version, so an upgrade lands under a new name.
         */
        fun copyBundledModel(context: Context): String? = try {
            val outFile = File(context.filesDir, MODEL_ASSET)
            if (!outFile.exists() || outFile.length() == 0L) {
                // Copy to a temp file then atomically rename, so a crash/kill/disk-full
                // mid-copy can't leave a partial file that exists() accepts on the next
                // launch (which would fail model load forever).
                val tempFile = File.createTempFile("model_", ".tmp", context.filesDir)
                try {
                    context.assets.open(MODEL_ASSET).use { input ->
                        tempFile.outputStream().use { output -> input.copyTo(output) }
                    }
                    if (!tempFile.renameTo(outFile)) {
                        throw java.io.IOException("failed to rename model temp file to $outFile")
                    }
                } catch (e: Exception) {
                    tempFile.delete()
                    throw e
                }
            }
            outFile.absolutePath
        } catch (e: Exception) {
            Log.e("QuailProcessor", "model copy failed", e)
            null
        }

        @JvmStatic private external fun nativeCreate(licenseKey: String, modelPath: String, outError: IntArray): Long
        @JvmStatic private external fun nativeProcess(handle: Long, input: ByteArray, len: Int): ByteArray?
        @JvmStatic private external fun nativeTakeError(handle: Long): Int
        @JvmStatic private external fun nativeSetEnabled(handle: Long, enabled: Boolean)
        @JvmStatic private external fun nativeReset(handle: Long)
        @JvmStatic private external fun nativeDestroy(handle: Long)
    }
}
